File: src/core/Account/UserState.cs
using System;
using System.Collections.Generic;
using core.Shared;
using Microsoft.FSharp.Core;

namespace core.Account
{
    public class UserState : IAggregateState
    {
        public Guid Id { get; private set; }
        public DateTimeOffset Created { get; private set; }
        public string Email { get; private set; }
        public string Firstname { get; private set; }
        public string Lastname { get; private set; }
        private string PasswordHash { get; set; }
        private string Salt { get; set; }
        public DateTimeOffset? Deleted { get; private set; }
        public string DeleteFeedback { get; private set; }
        public DateTimeOffset? Verified { get; private set; }
        public string SubscriptionLevel { get; private set; } = "Free";
        public bool IsPasswordAvailable => GetSalt() != null;
        public string Name => $"{Firstname} {Lastname}";
        public string BrokerageAccessToken { get; private set; }
        public string BrokerageRefreshToken { get; private set; }
        public DateTimeOffset BrokerageAccessTokenExpires { get; private set; }
        public DateTimeOffset BrokerageRefreshTokenExpires { get; private set; }
        public bool ConnectedToBrokerage { get; private set; }
        public bool BrokerageAccessTokenExpired => BrokerageAccessTokenExpires < DateTimeOffset.UtcNow;
        public FSharpOption<decimal> MaxLoss { get; private set; }
        private readonly HashSet<string> _brokerageTransactionIdSet = [];
        public decimal InterestReceived { get; private set; }
        public decimal CashTransferred { get; private set; }

        internal void ApplyInternal(UserCreated c)
        {
            Id = c.AggregateId;
            Created = c.When;
            Email = c.Email;
            Firstname = c.Firstname;
            Lastname = c.Lastname;
        }

        public void ApplyInternal(UserPasswordSet p)
        {
            PasswordHash = p.Hash;
            Salt = p.Salt;
        }

        [Obsolete("No longer tracking this")]
        internal void ApplyInternal(UserLoggedIn l)
        {
        }

        internal void ApplyInternal(UserDeleted d)
        {
            Deleted = d.When;
            DeleteFeedback = d.Feedback;
        }

        internal void ApplyInternal(UserConfirmed d)
        {
            Verified = d.When;
        }
        
        // TODO: comment out once this is moved to core.fs
        internal void ApplyInternal(UserSubscribedToPlan p)
        {
            // SubscriptionLevel = p.PlanId == Plans.Starter ? "Starter" : "Full";
        }

        internal void ApplyInternal(UserPasswordResetRequested _)
        {
            // no state to modify, this event gets captured via INotification mechanism
        }

        private void RefreshBrokerageData(string accessToken, string refreshToken, DateTimeOffset eventTimestamp)
        {
            ConnectedToBrokerage = true;
            BrokerageAccessToken = accessToken;
            BrokerageRefreshToken = refreshToken;
            BrokerageAccessTokenExpires = eventTimestamp.AddSeconds(1800); // these durations should come from somewhere else
            BrokerageRefreshTokenExpires = eventTimestamp.AddDays(7);
        }

        internal void ApplyInternal(UserConnectedToBrokerage e) =>
            RefreshBrokerageData(e.AccessToken, e.RefreshToken, e.When);
        
        internal void ApplyInternal(UserRefreshedBrokerageConnection e) =>
            RefreshBrokerageData(e.AccessToken, e.RefreshToken, e.When);

        internal void ApplyInternal(UserDisconnectedFromBrokerage _)
        {
            ConnectedToBrokerage = false;
            BrokerageAccessToken = null;
            BrokerageRefreshToken = null;
            BrokerageAccessTokenExpires = DateTimeOffset.MinValue;
        }

        private void ApplyInternal(UserBrokerageInterestApplied e)
        {
            _brokerageTransactionIdSet.Add(e.ActivityId);
            InterestReceived += e.NetAmount;
        }

        private void ApplyInternal(UserCashTransferApplied e)
        {
            _brokerageTransactionIdSet.Add(e.ActivityId);
            CashTransferred += e.NetAmount;
        }

        public bool ContainsBrokerageTransaction(string activityId)
        {
            return _brokerageTransactionIdSet.Contains(activityId);
        }

        internal void ApplyInternal(UserSettingSet e)
        {
            switch (e.Key)
            {
                case "maxLoss":
                    var maxLoss = decimal.Parse(e.Value);
                    MaxLoss = FSharpOption<decimal>.Some(maxLoss);
                    break;
                default:
                    throw new InvalidOperationException($"Unknown setting: {e.Key}");
            }
        }

        public bool PasswordHashMatches(string hash)
        {
            return PasswordHash == hash;
        }

        public string GetSalt()
        {
            return Salt;
        }

        public void Apply(AggregateEvent e)
        {
            ApplyInternal(e);
        }

        protected void ApplyInternal(dynamic obj)
        {
            ApplyInternal(obj);
        }
    }
}

File: src/core2.fs/Account/UserState.fs
namespace core.Account

open System
open System.Collections.Generic
open core.Shared
open Microsoft.FSharp.Core

type UserState() =
    let mutable _id = Guid.Empty
    let mutable _created = DateTimeOffset.MinValue
    let mutable _email = null
    let mutable _firstname = null
    let mutable _lastname = null
    let mutable _passwordHash = null
    let mutable _salt = null
    let mutable _deleted: DateTimeOffset option = None
    let mutable _deleteFeedback = null
    let mutable _verified: DateTimeOffset option = None
    let mutable _subscriptionLevel = "Free"
    let mutable _brokerageAccessToken = null
    let mutable _brokerageRefreshToken = null
    let mutable _brokerageAccessTokenExpires = DateTimeOffset.MinValue
    let mutable _brokerageRefreshTokenExpires = DateTimeOffset.MinValue
    let mutable _connectedToBrokerage = false
    let mutable _maxLoss: decimal option = None
    let _brokerageTransactionIdSet = HashSet<string>()
    let mutable _interestReceived = 0m
    let mutable _cashTransferred = 0m

    member _.Id with get() = _id
    member _.Created with get() = _created
    member _.Email with get() = _email
    member _.Firstname with get() = _firstname
    member _.Lastname with get() = _lastname
    member _.Deleted with get() = _deleted
    member _.DeleteFeedback with get() = _deleteFeedback
    member _.Verified with get() = _verified
    member _.SubscriptionLevel with get() = _subscriptionLevel
    member _.IsPasswordAvailable = _salt <> null
    member _.Name = sprintf "%s %s" _firstname _lastname
    member _.BrokerageAccessToken with get() = _brokerageAccessToken
    member _.BrokerageRefreshToken with get() = _brokerageRefreshToken
    member _.BrokerageAccessTokenExpires with get() = _brokerageAccessTokenExpires
    member _.BrokerageRefreshTokenExpires with get() = _brokerageRefreshTokenExpires
    member _.ConnectedToBrokerage with get() = _connectedToBrokerage
    member _.BrokerageAccessTokenExpired = _brokerageAccessTokenExpires < DateTimeOffset.UtcNow
    member _.MaxLoss with get() = _maxLoss
    member _.InterestReceived with get() = _interestReceived
    member _.CashTransferred with get() = _cashTransferred

    member internal this.ApplyInternal(c: UserCreated) =
        _id <- c.AggregateId
        _created <- c.When
        _email <- c.Email
        _firstname <- c.Firstname
        _lastname <- c.Lastname

    member this.ApplyInternal(p: UserPasswordSet) =
        _passwordHash <- p.Hash
        _salt <- p.Salt

    [<Obsolete("No longer tracking this")>]
    member internal this.ApplyInternal(l: UserLoggedIn) = ()

    member internal this.ApplyInternal(d: UserDeleted) =
        _deleted <- Some d.When
        _deleteFeedback <- d.Feedback

    member internal this.ApplyInternal(d: UserConfirmed) =
        _verified <- Some d.When

    member internal this.ApplyInternal(p: UserSubscribedToPlan) = ()

    member internal this.ApplyInternal(_: UserPasswordResetRequested) = ()

    member private this.RefreshBrokerageData(accessToken: string, refreshToken: string, eventTimestamp: DateTimeOffset) =
        _connectedToBrokerage <- true
        _brokerageAccessToken <- accessToken
        _brokerageRefreshToken <- refreshToken
        _brokerageAccessTokenExpires <- eventTimestamp.AddSeconds(1800.0)
        _brokerageRefreshTokenExpires <- eventTimestamp.AddDays(7.0)

    member internal this.ApplyInternal(e: UserConnectedToBrokerage) =
        this.RefreshBrokerageData(e.AccessToken, e.RefreshToken, e.When)

    member internal this.ApplyInternal(e: UserRefreshedBrokerageConnection) =
        this.RefreshBrokerageData(e.AccessToken, e.RefreshToken, e.When)

    member internal this.ApplyInternal(_: UserDisconnectedFromBrokerage) =
        _connectedToBrokerage <- false
        _brokerageAccessToken <- null
        _brokerageRefreshToken <- null
        _brokerageAccessTokenExpires <- DateTimeOffset.MinValue

    member private this.ApplyInternal(e: UserBrokerageInterestApplied) =
        _brokerageTransactionIdSet.Add(e.ActivityId) |> ignore
        _interestReceived <- _interestReceived + e.NetAmount

    member private this.ApplyInternal(e: UserCashTransferApplied) =
        _brokerageTransactionIdSet.Add(e.ActivityId) |> ignore
        _cashTransferred <- _cashTransferred + e.NetAmount

    member this.ContainsBrokerageTransaction(activityId: string) =
        _brokerageTransactionIdSet.Contains(activityId)

    member internal this.ApplyInternal(e: UserSettingSet) =
        match e.Key with
        | "maxLoss" ->
            let maxLoss = Decimal.Parse(e.Value)
            _maxLoss <- Some maxLoss
        | _ ->
            raise (InvalidOperationException(sprintf "Unknown setting: %s" e.Key))

    member this.PasswordHashMatches(hash: string) =
        _passwordHash = hash

    member this.GetSalt() = _salt

    member this.Apply(e: AggregateEvent) =
        this.ApplyInternal(e :> obj)

    member private this.ApplyInternal(obj: obj) =
        match obj with
        | :? UserCreated as e -> this.ApplyInternal(e)
        | :? UserPasswordSet as e -> this.ApplyInternal(e)
        | :? UserLoggedIn as e -> this.ApplyInternal(e)
        | :? UserDeleted as e -> this.ApplyInternal(e)
        | :? UserConfirmed as e -> this.ApplyInternal(e)
        | :? UserSubscribedToPlan as e -> this.ApplyInternal(e)
        | :? UserPasswordResetRequested as e -> this.ApplyInternal(e)
        | :? UserConnectedToBrokerage as e -> this.ApplyInternal(e)
        | :? UserRefreshedBrokerageConnection as e -> this.ApplyInternal(e)
        | :? UserDisconnectedFromBrokerage as e -> this.ApplyInternal(e)
        | :? UserBrokerageInterestApplied as e -> this.ApplyInternal(e)
        | :? UserCashTransferApplied as e -> this.ApplyInternal(e)
        | :? UserSettingSet as e -> this.ApplyInternal(e)
        | _ -> ()

    interface IAggregateState with
        member this.Id = this.Id
        member this.Apply(e) = this.Apply(e)
